app/forms/support_interface/multiple_provider_users_wizard.rb

```diff
@@ -0,0 +1,161 @@
+
+    attr_accessor :provider_id, :state_store, :provider_users, :stored_provider_users, :position
+    validates :provider_users, presence: true
+
+    def self.build(state_store:, provider_id:, position: nil)
+      state = state_store.read
+      stored_provider_users = JSON.parse(state)['provider_users'] if state
+      create_csv_row = ->(pu) { [pu['first_name'], pu['last_name'], pu['email_address']].compact.join(',') + "\n" }
+      provider_users = stored_provider_users&.map { |pu| create_csv_row.call(pu) }&.join
+
+      new(
+        state_store: state_store,
+        provider_id: provider_id,
+        stored_provider_users: stored_provider_users,
+        provider_users: provider_users,
+        position: position,
+      )
+    end
+
+    def clear_state!
+      @state_store.delete
+    end
+
+    def save_users_to_state_store!
+      state_store.write(prepare_provider_users_for_state_store)
+    end
+
+    def save_user_to_state_store!(single_provider_user_form)
+      state_store.write(prepare_provider_user_for_state_store(single_provider_user_form))
+    end
+
+    def save_to_db_and_invite!
+      prepare_data_for_db_save.each(&:call)
+    end
+
+    def provider_user_count
+      read_state['provider_users'].length
+    end
+
+    def position_and_count(state_store_id)
+      state = read_state
+      index = state['provider_users'].index do |provider_user|
+        [state_store_id].include?(provider_user['state_store_id'])
+      end
+
+      {
+        position: index + 1,
+        count: provider_user_count,
+      }
+    end
+
+    def state_store_id_for_position(position)
+      read_state['provider_users'][position - 1]['state_store_id']
+    end
+
+    def more_users_to_process?
+      read_state['provider_users'].select { |pu| pu['complete'] == false }.present?
+    end
+
+    def single_provider_user_form(state_store_id: nil)
+      provider_users = read_state['provider_users']
+      provider_user = if state_store_id
+                        provider_users.select { |pu| pu['state_store_id'] == state_store_id }.first
+                      else
+                        provider_users.select { |pu| pu['complete'] == false }&.first
+                      end
+
+      form = SupportInterface::CreateSingleProviderUserForm.new(
+        state_store_id: provider_user['state_store_id'],
+        provider_id: provider_id,
+        first_name: provider_user['first_name'],
+        last_name: provider_user['last_name'],
+        email_address: provider_user['email_address'],
+      )
+
+      if provider_user['permissions']
+        form.provider_permissions = { provider_permission: provider_user['permissions'].symbolize_keys }
+      end
+
+      form
+    end
+
+  private
+
+    def prepare_provider_users_for_state_store
+      provider_user_array_hash = provider_users
+        .split("\r\n")
+        .map { |row| row.split(/[\t,]+/) }
+        .map do |provider_user|
+        {
+          state_store_id: SecureRandom.uuid,
```

Comment:
can we use the position of the provider user as the key, instead of defining a `state_store_id` uuid?

Reply:
Yep that's possible. I used a uuid in order to future-proof the process. If for example, we decide later to be able to remove ProviderUsers part way through the flow (ie say a duplicate provider user is uploaded) then using 'position' will cause issues.

Reply:
Agreed to leave as is for now. May revisit if action on backlinks under taken

Reply:
We've agreed to use the position of the provider user as the key

Reply:
Fix - https://github.com/DFE-Digital/apply-for-teacher-training/pull/4896/commits/141db015de44bb17de767687e5ed76b014a18e0c